### PR TITLE
[MODINVOICE-93]-Approve invoice permission and require approval to mark as paid #93

### PR DESCRIPTION
**Pull request by yuntianhu**
## Purpose

https://issues.folio.org/browse/MODINVOICE-93
## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate action.

- Does this PR meet or exceed the expected quality standards?
  - [ ] Code coverage on new code is 80% or greater
  - [ ] Duplications on new code are 3% or less
  - [ ] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added, or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [ ] There are no breaking changes in this PR.

If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all the appropriate links to blocked/related issues?
- Are the JIRAs under active development?
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally, all the PRs involved in breaking changes would be merged on the same day
to avoid breaking the folio-testing environment.
Communication is paramount if that is to be achieved,
especially as the number of inter-module and inter-team dependencies increase.

While it's helpful for reviewers to help identify potential problems,
ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

File: src/main/java/org/folio/rest/impl/InvoiceHelper.java [hunk 1]
```diff
@@ -349,7 +352,9 @@ private Future<Void> validateAndHandleInvoiceStatusTransition(Invoice invoice, I
         validator.validateInvoice(invoice, invoiceFromStorage);
         verifyUserHasManagePermission(invoice.getAcqUnitIds(), invoiceFromStorage.getAcqUnitIds(), okapiHeaders);
         verifyUserHasFiscalYearUpdatePermission(invoice.getFiscalYearId(), invoiceFromStorage.getFiscalYearId(), okapiHeaders);
-
+        verifyUserHasInvoiceApprovePermission(invoice.getStatus(), invoiceFromStorage.getStatus(), okapiHeaders );
+        verifyUserHasInvoicePayPermission(invoice.getStatus(), invoiceFromStorage.getStatus(),okapiHeaders );
```
Comment (SerhiiNosko) [inline src/main/java/org/folio/rest/impl/InvoiceHelper.java:356]: small formatting issue, space before closing brackets can be deleted

File: src/main/java/org/folio/rest/impl/InvoiceHelper.java [hunk 2]
```diff
@@ -373,6 +378,8 @@ private Future<Void> validateAndHandleInvoiceStatusTransition(Invoice invoice, I
         .compose(lines -> updateEncumbranceLinksWhenFiscalYearIsChanged(invoice, invoiceFromStorage, invoiceLines)));
   }
 
+
+
```
Comment (SerhiiNosko) [inline src/main/java/org/folio/rest/impl/InvoiceHelper.java:382]: these new lines not needed here

File: src/main/java/org/folio/utils/UserPermissionsUtil.java [hunk 3]
```diff
@@ -63,12 +67,31 @@ public static void verifyUserHasManagePermission(List<String> newAcqUnitIds, Lis
     }
   }
 
+ public static void  verifyUserHasInvoicePayPermission  (Invoice.Status newInvoiceStatus, Invoice.Status statusFromStorage,   Map<String, String> okapiHeaders ){
```
Comment (SerhiiNosko) [inline src/main/java/org/folio/utils/UserPermissionsUtil.java:70]: many spaces, before opening brackets, also after statusFromStorage

File: src/main/java/org/folio/utils/UserPermissionsUtil.java [hunk 4]
```diff
@@ -63,12 +67,31 @@ public static void verifyUserHasManagePermission(List<String> newAcqUnitIds, Lis
...
+      if(!(newInvoiceStatus.toString().equals("Approved")) &&!(newInvoiceStatus.toString().equals("Paid"))&& isUserDoesNotHaveDesiredPermission(PAY, okapiHeaders)){
+      throw new HttpException(HttpStatus.HTTP_FORBIDDEN.toInt(),  CANNOT_PAY_INVOICE_WITHOUT_APPROVAL);
+    }
+ }
+ }
+
+  public static void verifyUserHasInvoiceApprovePermission (Invoice.Status newInvoiceStatus, Invoice.Status statusFromStorage, Map<String, String> okapiHeaders ){
+
+    if (isInvoiceStatusUpdated(newInvoiceStatus, statusFromStorage ) && isUserDoesNotHaveDesiredPermission(APPROVE, okapiHeaders)){
+
```
Comment (SerhiiNosko) [inline src/main/java/org/folio/utils/UserPermissionsUtil.java:82]: lets use the same code style, these empty new lines not needed here

File: src/main/java/org/folio/utils/UserPermissionsUtil.java [hunk 5]
```diff
@@ -63,12 +67,31 @@ public static void verifyUserHasManagePermission(List<String> newAcqUnitIds, Lis
     }
   }
 
+ public static void  verifyUserHasInvoicePayPermission  (Invoice.Status newInvoiceStatus, Invoice.Status statusFromStorage,   Map<String, String> okapiHeaders ){
+
+    if (isInvoiceStatusUpdated(newInvoiceStatus, statusFromStorage)){
+      if(!(newInvoiceStatus.toString().equals("Approved")) &&!(newInvoiceStatus.toString().equals("Paid"))&& isUserDoesNotHaveDesiredPermission(PAY, okapiHeaders)){
```
Comment (SerhiiNosko) [inline src/main/java/org/folio/utils/UserPermissionsUtil.java:73]: don't need to convert to strings, you can just compare enums with ==

Comment (SerhiiNosko) [reply src/main/java/org/folio/utils/UserPermissionsUtil.java:73]: formatting is incorrect in this line, see how another if statement written and use the same formatting

Comment (SerhiiNosko) [reply src/main/java/org/folio/utils/UserPermissionsUtil.java:73]: we should not check both permissions in this method.
We should have 2 separate methods, in one of them check Approve permission, in another - Pay.
These permissions are not connected between each other

File: src/main/java/org/folio/rest/impl/InvoiceHelper.java (hunk 2 above)
Comment (yuntianhu) [reply src/main/java/org/folio/rest/impl/InvoiceHelper.java:382]: will remove them

File: src/main/java/org/folio/utils/UserPermissionsUtil.java (hunk 3 above)
Comment (yuntianhu) [reply src/main/java/org/folio/utils/UserPermissionsUtil.java:70]: will remove those spaces

File: src/main/java/org/folio/utils/UserPermissionsUtil.java (hunk 4 above)
Comment (yuntianhu) [reply src/main/java/org/folio/utils/UserPermissionsUtil.java:82]: will remove the new lines

**Review by SerhiiNosko — APPROVED** [review]

**Review by azizbekxm — APPROVED** [review]

**Pull request merged**